Rewrite scraper using classes

ondenman opened the pull request:
Source data tables have changed layout and so some members are not
being scraped. The scraper has been rewritten to address this issue.

ondenman committed:
Update Ruby version to 2.3.0

ondenman committed:
Add require_all gem

ondenman committed:
Remove table scraping logic

Removes scraping logic from scraper.rb. This is a step towards
implementing it in separate classes.

ondenman committed:
Create Table class

Table represents a table on the web page. It will be responsible for
returning a list of Row objects. WIP

ondenman committed:
Create Row class

WIP Represents the data stored in a table row.

ondenman committed:
Add field serializer gem

ondenman committed:
Skip row when it does not contain member info

The tables have a single column row at the top which does not contain member information. We skip these rows in the map loop -- the item becomes nil. We then use compact to remove the nil items.

ondenman committed:
Add fields to Row

:id
:name
:coalition
:coalition_id
:party
:party_id
:state
:wikipedia
:wikipedia_en
:start_date
:end_date

ondenman committed:
Rename Table#rows to Table#to_h

ondenman committed:
Swap constituency and constituency_id

Variables were being assigned the wrong values. Swaapping them around corrects this.

ondenman committed:
Handle rows without coalitions

Some members are not affiliated with a coalition. In these cases
the party is listed in the position of the coalition. This change
makes sure the party data is stored in the correct field.

ondenman committed:
Change to_h to to_a

Reflects that a table is a list of members

ondenman committed:
Add Term class

Represents a single wikipedia term page

ondenman committed:
Add Page class

Copied from Austria scraper

ondenman committed:
Make Term subclass Page

ondenman committed:
Replace to_h with field serializer method

Uses field serializer to generate the hash that contains the list of members.

ondenman committed:
Handle instances where row contains start and end date

Where the start or end date of a member differs from the start and end of a term, the alternative dates are listed in the row. This commit handles any occurrence where both an alternative start and end are given. (Previously, we could only handle occurences of a single alternative date and the scraper would crash if both dates were listed.)

ondenman committed:
Add Row#affiliation_cell

Moves duplicate code into its own method.

ondenman committed:
Try to extract year given a badly formed date

In the event that Date#parse raises an exception, we try to extract
the year from the string.

For example:

```
date_or_year("1968") #=> "1968"
date_or_year("12 December 1967") #=> 1967-12-12"
```

ondenman committed:
Use extracted classes in scrape_term

Scraper has been rewritten using classes. Scrape_term now instantiates
Term objects.

ondenman committed:
Fix incorrect use of name instead of method

ondenman committed:
Return nil if no party_id is listed in row

No longer returns "unknown".

ondenman committed:
Return nil if no party is listed in row

No longer returns "unknown".

ondenman committed:
Return nil if no coalition_id is listed in row

No longer returns "unknown".

ondenman committed:
Return nil if no coalition is listed in row

No longer returns "unknown".

tmtmtmtm commented on Gemfile:
```diff
@@ -4,11 +4,14 @@
 source "https://rubygems.org"
 
-ruby "2.0.0"
+ruby "2.3.0"
```
Interesting choice…  Any reason for not going to 2.3.1?

tmtmtmtm commented on lib/term.rb:
```diff
@@ -0,0 +1,17 @@
+require 'nokogiri'
+require_relative 'table'
+
+class Term < Page
```
As some of these classes are Pages, and some are just parts, it might be clearer to be more explicit with this one and call it TermPage

ondenman commented:
Scraper update is a step towards fixing: https://github.com/everypolitician/everypolitician-data/pull/18469

tmtmtmtm commented on lib/term.rb:
```diff
@@ -0,0 +1,17 @@
+
+  field :members do
+    tables.map do |table|
+      Table.new(table).to_a
```
There's a slight semantic confusion with naming here, where you have a `Table` class, but `table` is referring to something that isn't a `Table`. I'd suggest calling this inner `table`, something more like `node` to make it clearer that it's simply a nokogiri Node, not an actual Table.

tmtmtmtm commented on lib/term.rb:
```diff
@@ -0,0 +1,17 @@
+  field :members do
+    tables.map do |table|
+      Table.new(table).to_a
+    end.flatten
```
`.map { … }.flatten` is usually cleaner as `.flat_map`

(and would let you collapse this into `tables.flat_map { |node| Table.new(node).to_a }`

tmtmtmtm commented on lib/table.rb:
```diff
@@ -0,0 +1,23 @@
+    @table = node
+  end
+
+  def to_a
```
This is quite a complicated method. Parts of this should almost certainly be split out into individual methods. I think it would probably also much clearer if it followed the same pattern of using field_serializer rather than taking it own different approach.

tmtmtmtm commented on lib/table.rb:
```diff
@@ -0,0 +1,23 @@
+      tds = tr.xpath('./td')
+      next if tds.count == 1
+      constituency_id = tds.shift.text.strip if tds.count > 2
+      constituency = tds.shift.text.strip if tds.count > 2
```
Rather than having all this logic in-place, which is likely to just grow more and more complex over time, I think this would be much simpler and much clearer with separate Table classes, each representing a different layout.

tmtmtmtm commented on lib/row.rb:
```diff
@@ -0,0 +1,125 @@
+          .downcase
+          .gsub(/ /,'_')
+          .gsub('_(page_does_not_exist)','')
+  end
```
Is every person guaranteed to have a link? It's fairly common on Wikipedia for a person to simply be plain text, and not linked.

tmtmtmtm commented on lib/row.rb:
```diff
@@ -0,0 +1,125 @@
+  field :state do
+    tds.xpath('./preceding::h3[1]')
+       .text
+       .gsub('[edit]', '')
```
This seems quite out of place. I think you want to descend into the span that holds the text, rather than taking the entire h3

tmtmtmtm commented on lib/row.rb:
```diff
@@ -0,0 +1,125 @@
+  end
+
+  field :wikipedia do
+    wiki_link
```
Why do you need a separate method for this, rather than just having that method be this field?

tmtmtmtm commented on lib/row.rb:
```diff
@@ -0,0 +1,125 @@
+  end
+
+  field :wikipedia_en do
+    wiki_name
```
Ditto here.

tmtmtmtm reviewed (changes requested):
This could probably do with having rubocop run against it too…

ondenman committed:
Update Ruby version to 2.3.1

ondenman committed:
Rename Term as TermPage

To make it explicit that this Class represents a type of page.

ondenman committed:
Rename table variable to avoid semantic confusion

Renamed the `table` variable within the tables.map loop to avoid
any semantic confusion with `Table#new` or the object it returns.

ondenman committed:
Replace #map with #flat_map

Simplifies the method.

ondenman committed:
Create member id from the element text when name is not within an anchor tag

It is not guaranteed that every member has a link. Previously, we created
ids from the anchor's title attribute. Now, if this is not available
we create the id from the text value of the element.

ondenman committed:
Remove #wiki_link

Moves the #wiki_link code into the wikipedia field block.

ondenman committed:
Remove #wiki_name

Moves the #wiki_name code into the :wikipedia_en field block.

ondenman committed:
Add rubocop

ondenman committed:
Drill down to specific <span> element to capture name of State

Previously, we were capturing the entire text content of the
containing <h3> and substituting out unwanted text.

This change removes the need for a gsub as it only captures the
text belonging to the specific span that contains the State name.

ondenman committed:
Reference table cells by negative indices

Rows in table can have extra cells at the beginning.

Indexing from the right provides a consistent way to access
the cells.

ondenman committed:
Handle edge case where member name is listed as 'vacant'

A table row on the term 12 page list the member name as
'vacant'. Since we don't wan a member called 'vacant', we
do not return data for this row.

ondenman committed:
Rename Table to Members

New name reflects purpose of class's purpose: to represent
a list of members.

ondenman committed:
Extract #member from loop in #to_a

Refactoring.

Rename #member_row to #member

ondenman committed:
Remove handling of member named 'vacant' from Row

Removing responsibility of handling case where a row
represents a vacant seat. Row shouldn't care as its job
is just to fetch the data.

ondenman committed:
fixup! Remove handling of member named 'vacant' from Row

ondenman committed:
Replace tds[-2] with #name_cell

Replaces multiple references to tds[-2] with single method.

ondenman committed:
Handle 'vacant' seats in Members

Members is responsible for returning a list of members so should
not return any vacant seat data.

tmtmtmtm commented:
superseded
